lib/extras/encode_jpeg.cc

```diff
@@ -0,0 +1,340 @@
+    offset[i] = offset[i - 1] + huff->counts[i - 1];
+  }
+  for (size_t i = 0; i <= jpeg::kJpegHuffmanAlphabetSize; ++i) {
+    if (depth[i] > 0) {
+      huff->values[offset[depth[i]]++] = i;
+    }
+  }
+  huff->is_last = false;
+}
+
+void AddJpegHuffmanCodes(std::vector<Histogram>& histograms,
+                         size_t slot_id_offset,
+                         std::vector<jpeg::JPEGHuffmanCode>* huff_codes) {
+  for (size_t i = 0; i < histograms.size(); ++i) {
+    jpeg::JPEGHuffmanCode huff_code;
+    huff_code.slot_id = slot_id_offset + i;
+    std::vector<uint32_t> counts(jpeg::kJpegHuffmanAlphabetSize + 1);
+    std::vector<uint8_t> depths(jpeg::kJpegHuffmanAlphabetSize + 1);
+    for (size_t j = 0; j < jpeg::kJpegHuffmanAlphabetSize; ++j) {
+      counts[j] = histograms[i].data_[j];
+    }
+    counts[jpeg::kJpegHuffmanAlphabetSize] = 1;
+    CreateHuffmanTree(counts.data(), counts.size(),
+                      jpeg::kJpegHuffmanMaxBitLength, &depths[0]);
+    BuildJpegHuffmanCode(&depths[0], &huff_code);
+    huff_codes->emplace_back(std::move(huff_code));
+  }
+}
+
+void FillJPEGData(const Image3F& opsin, const ImageF& qf, float dc_quant,
+                  const FrameDimensions& frame_dim, jpeg::JPEGData* out) {
+  // ICC
+  out->marker_order.push_back(0xe2);
+  out->app_data.push_back(CreateXybICCAppMarker());
+
+  // DQT
+  out->marker_order.emplace_back(0xdb);
+  float qm[3 * kDCTBlockSize];
+  AddJpegQuantMatrices(qf, dc_quant, &out->quant, qm);
+
+  // SOF
+  out->marker_order.emplace_back(0xc2);
+  out->components.resize(3);
+  out->height = frame_dim.ysize;
+  out->width = frame_dim.xsize;
+  out->components[0].id = 'R';
+  out->components[1].id = 'G';
+  out->components[2].id = 'B';
+  for (size_t c = 0; c < 3; ++c) {
+    out->components[c].h_samp_factor = 1;
+    out->components[c].v_samp_factor = 1;
+    out->components[c].width_in_blocks = frame_dim.xsize_blocks;
+    out->components[c].height_in_blocks = frame_dim.ysize_blocks;
+    out->components[c].quant_idx = c;
+  }
+  HWY_DYNAMIC_DISPATCH(ComputeDCTCoefficients)
+  (opsin, qf, frame_dim, qm, &out->components);
+
+  // DHT (the actual Huffman codes will be added later).
+  out->marker_order.emplace_back(0xc4);
+
+  // SOS
+  std::vector<ProgressiveScan> progressive_mode = {
+      // DC
+      {0, 0, 0, 0, true},
+      // AC 1 - highest bits
+      {1, 63, 0, 1, false},
+      // AC 2 - lowest bit
+      {1, 63, 1, 0, false},
+  };
+  AddJpegScanInfos(progressive_mode, &out->scan_info);
+  for (size_t i = 0; i < out->scan_info.size(); i++) {
+    out->marker_order.emplace_back(0xda);
+  }
+
+  // EOI
+  out->marker_order.push_back(0xd9);
+
+  // Gather histograms.
+  jpeg::SerializationState ss;
+  JXL_CHECK(jpeg::ProcessJpeg(*out, &ss));
+
+  // Build DC Huffman codes and add them to DHT segment.
+  std::vector<Histogram> dc_histo;
+  std::vector<uint32_t> dc_tbl_indexes;
+  ClusterJpegHistograms(ss.dc_huff_table, 4, &dc_histo, &dc_tbl_indexes);
```

Comment:
I have to admit it's entirely unclear to me what is going on here:
- who sets ss.dc_huff_table?
- why are we clustering to 4 tables and assigning clusters to scans sequentially? On the top of my head I don't see any reason why this should work.

Reply:
Ah, nevermind, I didn't see the changes in jpeg_serialization_state, now it makes a lot more sense.

It's still probably not the optimal thing to do (in particular because clustering sort of assumes ANS, and ignores Huffman-specific things), but good enough for now.

Reply:
Agreed, added a TODO to use JPEG-specific clustering.